reactor-core/src/test/java/reactor/core/publisher/FluxUsingWhenTest.java

```diff
@@ -852,6 +855,180 @@ public void contextPropagationOnCancelWithNoHandler(Flux<String> source) {
+		LongAdder cleanupCount = new LongAdder();
+		Flux<String> flux = Flux.usingWhen(Mono.defer(() -> Mono.just("foo")), Mono::just,
+				s -> Mono.fromRunnable(() -> cleanupCount.add(10)), //10 for completion
+				s -> Mono.fromRunnable(() -> cleanupCount.add(100)), //100 for error
+				s -> Mono.fromRunnable(() -> cleanupCount.add(1000)) //1000 for cancel
+		);
+
+		flux.subscribe(new CoreSubscriber<Object>() {
+			Subscription subscription;
+
+			@Override
+			public void onSubscribe(Subscription s) {
+				s.request(1);
+				subscription = s;
+			}
+
+			@Override
+			public void onNext(Object o) {}
+
+			@Override
+			public void onError(Throwable t) {}
+
+			@Override
+			public void onComplete() {
+				subscription.cancel();
+			}
+		});
+
+		assertThat(cleanupCount.sum()).isEqualTo(10);
+	}
+
+	@Test
+	public void noCancelCallbackAfterError() {
+		LongAdder cleanupCount = new LongAdder();
+		Flux<String> flux = Flux.usingWhen(Mono.just("foo"), v -> Mono.error(new IllegalStateException("boom")),
+				s -> Mono.fromRunnable(() -> cleanupCount.add(10)), //10 for completion
+				s -> Mono.fromRunnable(() -> cleanupCount.add(100)), //100 for error
+				s -> Mono.fromRunnable(() -> cleanupCount.add(1000)) //1000 for cancel
+		);
+
+		flux.subscribe(new CoreSubscriber<Object>() {
+			Subscription subscription;
+
+			@Override
+			public void onSubscribe(Subscription s) {
+				s.request(1);
+				subscription = s;
+			}
+
+			@Override
+			public void onNext(Object o) {}
+
+			@Override
+			public void onError(Throwable t) {
+				subscription.cancel();
+			}
+
+			@Override
+			public void onComplete() {}
+		});
+
+		assertThat(cleanupCount.sum()).isEqualTo(100);
+	}
+
+	@Test
+	public void noCompleteCallbackAfterCancel() throws InterruptedException {
+		AtomicBoolean cancelled = new AtomicBoolean();
+		LongAdder cleanupCount = new LongAdder();
+
+		Publisher<String> badPublisher = s -> s.onSubscribe(new Subscription() {
+			@Override
+			public void request(long n) {
+				new Thread(() -> {
+					s.onNext("foo1");
+					try { Thread.sleep(100); } catch (InterruptedException e) {}
+					s.onComplete();
+				}).start();
+			}
+
+			@Override
+			public void cancel() {
+				cancelled.set(true);
+			}
+		});
+
+		Flux<String> flux = Flux.usingWhen(Mono.just("foo"), v -> badPublisher,
+				s -> Mono.fromRunnable(() -> cleanupCount.add(10)).log(), //10 for completion
+				s -> Mono.fromRunnable(() -> cleanupCount.add(100)), //100 for error
+				s -> Mono.fromRunnable(() -> cleanupCount.add(1000)) //1000 for cancel
+		);
+
+		flux.subscribe(new CoreSubscriber<String>() {
+			Subscription subscription;
+
+			@Override
+			public void onSubscribe(Subscription s) {
+				s.request(1);
+				subscription = s;
+			}
+
+			@Override
+			public void onNext(String o) {
+				subscription.cancel();
+			}
+
+			@Override
+			public void onError(Throwable t) {}
+
+			@Override
+			public void onComplete() {}
+		});
+
+		Thread.sleep(300);
```

Comment:
`await().untilAsserted`?

Reply:
nope, tried with awaitility but there is an intermediate state where the cancellation increment was performed, but not the complete increment (yet). awaitility would happily say the test was green.
Similarly, I couldn't make use of `atLeast` because if the assertion is true too early, awaitility complains.
So most straightforward way of checking the state is waiting for a bunch of ms and asserting.

Reply:
> where the cancellation increment was performed, but not the complete increment (yet)
We could assert this condition too so that it passes only when both are matched

Reply:
My intention was to avoid static sleeps because they add a constant time to the test execution time, but okay to not use it if it makes it harder to detect the error

Reply:
yeah that is indeed something I tried to avoid by trying with awaitility first, without success ;)